extensions/notification/NotificationQueuePopulator.js

```diff
@@ -119,22 +119,76 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
         return this.bnConfigManager.removeConfig(bucketName);
     }
 
+    /**
+     * Returns the correct versionId
+     * to display according to the
+     * versioning state of the object
+     * @param {Object} value log entry object
+     * @param {Object} overheadFields - extra fields missing from the log entry
+     * @return {String} versionId
+     */
+    _getVersionId(value, overheadFields) {
+        const versionId = value.versionId || (overheadFields && overheadFields.versionId);
+        const isNullVersion = value.isNull;
+        const isVersioned = !!versionId;
+        // Versioning suspended objects have
+        // a versionId, however it is internal
+        // and should not be used to get the object
+        if (isNullVersion || !isVersioned) {
+            return null;
+        }
+        return versionId;
+    }
+
+    /**
+     * Decides if we should process the entry.
+     * Since we get both master and version events,
+     * we need to avoid pushing two notifications for
+     * the same event.
+     * - For non-versioned buckets, we process the master
+     * objects' events.
+     * - For versioned buckets, we process version events
+     * and ignore all master events.
+     * - For versioning suspended buckets, we need to process
+     * both master and version events, as the master is considered
+     * a separate version.
+     * @param {String} key object key
+     * @param {Object} value object metadata
+     * @return {boolean} - true if entry is valid
+     */
+    _shouldProcessEntry(key, value) {
+        const isMaster = isMasterKey(key);
+        const hasVersionId = !!value.versionId;
+        const keyHasNullVersion = key.endsWith('\0');
+        const isVersionedMaster = isMaster && hasVersionId;
+        const isNullVersion = (isMaster || keyHasNullVersion) && value.isNull;
+        const isNullWithVersionId = hasVersionId && value.isNull && !keyHasNullVersion;
+
+        return (!isMaster || !isVersionedMaster || isNullVersion) && !isNullWithVersionId;
```

Comment:
since `isVersionedMaster` implies `isMaster` I think we can get rid of the `!isMaster ||` to simplify.

I think the logic is right, but it looks to me a bit convoluted as is. Also some comments would help understanding. Here's a suggestion to change the test while maintaining the same truth condition, and get rid of some of the intermediate variables (it's arguably a bit longer to read but I think simplifies the logic overall):

```
if (keyHasNullVersion) { // could be named 'isNullKey'
  return true; // null keys always generate a notification because they are never duplicated with a versioned key or a master key
}
if (!isMaster) {
  return !version.isNull; // versioned keys (other than null keys) representing a null version never generate a notification because they always duplicate the reference null-versioned master key. Other, regular (non-null) versioned keys generate a notification.
}
return !hasVersionId || version.isNull; // generate a notification for non-versioned or null-versioned master keys, but not for regular versions as then the versioned key triggers the notification.

Reply:
Actually I just realized that we also need to generate notifications for non-current null versions, which are normally stored in versioned keys in V0. If we don't have the context of the master key when we process a versioned key, we cannot know if it's a duplicate or not.

Is it possible to process versioned keys that have `isNull` knowing the context of the master? For example, knowing that it was updated atomically with the same value could allow to ignore the versioned update.

Without this context, the best we can do is always generate notifications for `isNull` versioned keys, leading to potential duplicates with the master. (it's better than skipping notifications).

Reply:
So this would be the new logic (I think) without the context of master updates:

```
if (!isMaster) {
  return true; // versioned keys do generate a notifications. FIXME: in some cases they may duplicate the notification with a master update of the same null version.
}
return !hasVersionId || version.isNull; // generate a notification for non-versioned or null-versioned master keys, but not for regular versions as then the versioned key triggers the notification.
```